File: app/buildz/tmp/kapt3/stubs/debug/com/bits/bee/bpmc/data/repository/CcTypeRepositoryImpl.java
package com.bits.bee.bpmc.data.repository;

import java.lang.System;

/**
 * Created by aldi on 30/03/22.
 */
@kotlin.Metadata(mv = {1, 7, 1}, k = 1, d1 = {"\u00000\n\u0002\u0018\u0002\n\u0002\u0018\u0002\n\u0000\n\u0002\u0018\u0002\n\u0000\n\u0002\u0018\u0002\n\u0000\n\u0002\u0018\u0002\n\u0002\b\u0002\n\u0002\u0018\u0002\n\u0002\u0018\u0002\n\u0002\u0010 \n\u0002\u0018\u0002\n\u0000\u0018\u00002\u00020\u0001B\u001f\b\u0007\u0012\u0006\u0010\u0002\u001a\u00020\u0003\u0012\u0006\u0010\u0004\u001a\u00020\u0005\u0012\u0006\u0010\u0006\u001a\u00020\u0007\u00a2\u0006\u0002\u0010\bJ\u001a\u0010\t\u001a\u0014\u0012\u0010\u0012\u000e\u0012\n\u0012\b\u0012\u0004\u0012\u00020\r0\f0\u000b0\nH\u0016R\u000e\u0010\u0002\u001a\u00020\u0003X\u0082\u0004\u00a2\u0006\u0002\n\u0000R\u000e\u0010\u0004\u001a\u00020\u0005X\u0082\u0004\u00a2\u0006\u0002\n\u0000R\u000e\u0010\u0006\u001a\u00020\u0007X\u0082\u0004\u00a2\u0006\u0002\n\u0000\u00a8\u0006\u000e"}, d2 = {"Lcom/bits/bee/bpmc/data/repository/CcTypeRepositoryImpl;", "Lcom/bits/bee/bpmc/domain/repository/CcTypeRepository;", "apiUtils", "Lcom/bits/bee/bpmc/data/data_source/remote/ApiUtils;", "ccTypeDao", "Lcom/bits/bee/bpmc/data/data_source/local/dao/CcTypeDao;", "ioDispatcher", "Lkotlinx/coroutines/CoroutineDispatcher;", "(Lcom/bits/bee/bpmc/data/data_source/remote/ApiUtils;Lcom/bits/bee/bpmc/data/data_source/local/dao/CcTypeDao;Lkotlinx/coroutines/CoroutineDispatcher;)V", "getLatestCcTypeList", "Lkotlinx/coroutines/flow/Flow;", "Lcom/bits/bee/bpmc/utils/Resource;", "", "Lcom/bits/bee/bpmc/domain/model/CcType;", "app_debug"})
public final class CcTypeRepositoryImpl implements com.bits.bee.bpmc.domain.repository.CcTypeRepository {
    private final com.bits.bee.bpmc.data.data_source.remote.ApiUtils apiUtils = null;
    private final com.bits.bee.bpmc.data.data_source.local.dao.CcTypeDao ccTypeDao = null;
    private final kotlinx.coroutines.CoroutineDispatcher ioDispatcher = null;
    
    @javax.inject.Inject()
    public CcTypeRepositoryImpl(@org.jetbrains.annotations.NotNull()
    com.bits.bee.bpmc.data.data_source.remote.ApiUtils apiUtils, @org.jetbrains.annotations.NotNull()
    com.bits.bee.bpmc.data.data_source.local.dao.CcTypeDao ccTypeDao, @org.jetbrains.annotations.NotNull()
    kotlinx.coroutines.CoroutineDispatcher ioDispatcher) {
        super();
    }
    
    @org.jetbrains.annotations.NotNull()
    @java.lang.Override()
    public kotlinx.coroutines.flow.Flow<com.bits.bee.bpmc.utils.Resource<java.util.List<com.bits.bee.bpmc.domain.model.CcType>>> getLatestCcTypeList() {
        return null;
    }
}

File: app/src/main/java/com/bits/bee/bpmc/data/repository/CcTypeRepositoryImpl.kt
package com.bits.bee.bpmc.data.repository

import com.bits.bee.bpmc.data.data_source.local.dao.CcTypeDao
import com.bits.bee.bpmc.data.data_source.remote.ApiUtils
import com.bits.bee.bpmc.data.data_source.remote.response.CcTypeResponse
import com.bits.bee.bpmc.domain.mapper.CcTypeDataMapper
import com.bits.bee.bpmc.domain.model.CcType
import com.bits.bee.bpmc.domain.repository.CcTypeRepository
import com.bits.bee.bpmc.utils.ApiResponse
import com.bits.bee.bpmc.utils.NetworkDatabaseBoundResource
import com.bits.bee.bpmc.utils.Resource
import kotlinx.coroutines.CoroutineDispatcher
import kotlinx.coroutines.flow.Flow
import javax.inject.Inject

/**
 * Created by aldi on 30/03/22.
 */
class CcTypeRepositoryImpl @Inject constructor(
    private val apiUtils: ApiUtils,
    private val ccTypeDao: CcTypeDao,
    private val ioDispatcher: CoroutineDispatcher
) : CcTypeRepository {

    override fun getLatestCcTypeList(): Flow<Resource<List<CcType>>> {
        return object : NetworkDatabaseBoundResource<List<CcType>, CcTypeResponse>(){
            override suspend fun loadFormDB(): List<CcType> {
                return ccTypeDao.getCcTypeList().map { CcTypeDataMapper.fromDbToDomain(it) }
            }

            override fun shouldFetch(data: List<CcType>?): Boolean {
                return true
            }

            override suspend fun createCall(): Flow<ApiResponse<CcTypeResponse>> {
                return apiUtils.getCcTypeApiService().getCcType()
            }

            override suspend fun saveCallResult(data: CcTypeResponse) {
                ccTypeDao.insertBulk(data.data.map { CcTypeDataMapper.fromNetworkToDb(it) })
            }
        }.getAsFlow()
    }
}